src/main/java/org/eclipsefoundation/marketplace/model/QueryParams.java

```diff
@@ -0,0 +1,142 @@
+		Objects.requireNonNull(params);
+
+		this.params = new HashMap<>(params);
+	}
+
+	/**
+	 * Retrieves the first value set in a list from the map for a given key.
+	 * 
+	 * @param params the parameter map containing the value
+	 * @param key    the key to retrieve the value for
+	 * @return the first value set in the parameter map for the given key, or null
+	 *         if absent.
+	 */
+	public String getFirstParam(String key) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		List<String> vals = this.params.get(key);
+		if (vals == null || vals.isEmpty()) {
+			return null;
+		}
+		return vals.get(0);
+	}
+
+	/**
+	 * Retrieves the value list from the map for a given key.
+	 * 
+	 * @param params the parameter map containing the values
+	 * @param key    the key to retrieve the values for
+	 * @return the value list for the given key if it exists, or an empty collection
+	 *         if none exists.
+	 */
+	public List<String> getParams(String key) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		List<String> vals = this.params.get(key);
+		if (vals == null || vals.isEmpty()) {
+			return Collections.emptyList();
+		}
+		return vals;
+	}
+
+	/**
+	 * <p>
+	 * Sets the given value for the given key, overriding existing values. Uses a
+	 * small load factor for the array to keep the memory footprint smaller, but the
+	 * object churn reasonable should more items need to be added.
+	 * </p>
+	 * <p>
+	 * A null value can be passed to this method for the value, which will remove
+	 * the current list values for the key if present.
+	 * </p>
+	 * 
+	 * @param params map containing parameters to update
+	 * @param key    string key to set the value for, must not be null
+	 * @param value  the value to set for the key
+	 */
+	public void setParam(String key, String value) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		// if null value, unset the key
+		if (value == null) {
+			this.params.remove(key);
+			return;
+		}
+
+		// if value exists, set as new map
+		List<String> param = new ArrayList<>(4);
+		param.add(value);
+		this.params.put(key, param);
+	}
+
+	/**
+	 * Adds the given value for the given key, preserving previous values if they
+	 * exist.
+	 * 
+	 * @param params map containing parameters to update
+	 * @param key    string key to add the value to, must not be null
+	 * @param value  the value to add to the key
+	 */
+	public void addParam(String key, String value) {
+		if (StringUtils.isBlank(key)) {
+			throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
+		}
+
+		// return immediately if theres no value
+		if (value == null) {
+			return;
+		}
+
+		// set the param, creating list if it doesn't exist
+		List<String> param = this.params.get(key);
+		if (param == null) {
+			this.setParam(key, value);
+		} else {
+			param.add(value);
+		}
+	}
```

Comment:
while reading addParam and setParam, I've the feeling that they should be merged into somehting like

    public void addParam(String key, String value) {
        if (StringUtils.isBlank(key)) {
             throw new IllegalArgumentException(EMPTY_KEY_MESSAGE);
         }
        if (value == null) {
            throw new NullPointerException("value must not be null");
        }
        map.computeIfAbsent(key, k -> new ArrayList<>()).add(value);
    }

and add the `unsetParam(String key)` method as mentioned above.